reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1227 @@
+	return (s + Sequence(t)) & uint32Max
+}
+
+type TCPAssemblyStats struct {
+	// For this ScatterGather
+	NbChunks  uint
+	NbPackets uint
+	// For the half connection, since last call to ReassembledSG()
+	NbQueuedBytes    uint
+	NbQueuedPackets  uint
+	NbOverlapBytes   uint
+	NbOverlapPackets uint
+}
+
+type ScatterGather interface {
+	// Returns the length of available bytes and saved bytes
+	Lengths() (int, int)
+	// Returns the bytes up to length
+	Fetch(length int) []byte
+	// Tell to keep from offset
+	KeepFrom(offset int)
+	// Return CaptureInfo of packet corresponding to given offset
+	CaptureInfo(offset int) gopacket.CaptureInfo
+	// Return some info about the reassembled chunks
+	Info() (direction TcpFlowDirection, start bool, end bool, skip int)
+	// Return some stats regarding the state of the stream
+	Stats() TCPAssemblyStats
+}
+
+// Containter: either page or LivePacket
+type ByteContainer interface {
+	Bytes() []byte
+	Length() int
+	ConvertToPages(*pageCache, int, AssemblerContext) (*page, *page, int)
+	CaptureInfo() gopacket.CaptureInfo
+	AssemblerContext() AssemblerContext
+	Release(*pageCache) int
+	Start() bool
+	End() bool
+	Seq() Sequence
+	IsPacket() bool
+}
+
+// Implements a ScatterGather
+type ReassemblyObject struct {
+	all       []ByteContainer
+	Skip      int
+	Direction TcpFlowDirection
+	saved     int
+	to_keep   int
+	// stats
+	nbQueuedBytes    uint
+	nbQueuedPackets  uint
+	nbOverlapBytes   uint
+	nbOverlapPackets uint
+}
+
+func (rl *ReassemblyObject) Lengths() (int, int) {
+	l := 0
+	for _, r := range rl.all {
+		l += r.Length()
+	}
+	return l, rl.saved
+}
+
+func (rl *ReassemblyObject) Fetch(l int) []byte {
+	if len(rl.all) == 1 && l == rl.all[0].Length() {
+		return rl.all[0].Bytes()
+	}
+	bytes := []byte{}
+	for _, bc := range rl.all {
+		bytes = append(bytes, bc.Bytes()...)
+	}
+	return bytes[:l]
+}
+
+func (rl *ReassemblyObject) KeepFrom(offset int) {
+	rl.to_keep = offset
+}
+
+func (rl *ReassemblyObject) CaptureInfo(offset int) gopacket.CaptureInfo {
+	current := 0
+	for _, r := range rl.all {
+		if current >= offset {
+			return r.CaptureInfo()
+		}
+		current += r.Length()
+	}
+	// Invalid offset
+	return gopacket.CaptureInfo{}
+}
+
+func (rl *ReassemblyObject) Info() (TcpFlowDirection, bool, bool, int) {
+	return rl.Direction, rl.all[0].Start(), rl.all[len(rl.all)-1].End(), rl.Skip
+}
+
+func (rl *ReassemblyObject) Stats() TCPAssemblyStats {
+	nbPackets := uint(0)
+	for _, r := range rl.all {
+		if r.IsPacket() {
+			nbPackets++
+		}
+	}
+	return TCPAssemblyStats{
+		NbChunks:         uint(len(rl.all)),
+		NbPackets:        nbPackets,
+		NbQueuedBytes:    rl.nbQueuedBytes,
+		NbQueuedPackets:  rl.nbQueuedPackets,
+		NbOverlapBytes:   rl.nbOverlapBytes,
+		NbOverlapPackets: rl.nbOverlapPackets,
+	}
+}
+
+const pageBytes = 1900
+
+// TCP flow direction
+type TcpFlowDirection bool
+
+const (
+	TcpDirServerToClient TcpFlowDirection = false
+	TcpDirClientToServer TcpFlowDirection = true
```

Comment:
what happens if direction is unknown?  how does reassembly handle streams it starts seeing halfway through?